dali/base/dasds.cpp

```diff
@@ -3675,6 +3790,78 @@ class CLockInfo : public CInterface, implements IInterface
         return out;
     }
 
+    bool getFilteredLocks(CLockFilters &filters, MemoryBuffer &mb)
+    {
+        bool isFile;
+        StringBuffer file;
+        if (!filters.checkAndReadPath(xpath, isFile, file))
+            return false;
+
+        unsigned nlocks=0;
+        MemoryBuffer locks, localMB;
+        UInt64Array keys;
+        {
+            CHECKEDCRITICALBLOCK(crit, fakeCritTimeout);
+            HashIterator iter(connectionInfo);
+            ForEach(iter)
+            {
+                IMapping &imap = iter.query();
+                LockData *lD = connectionInfo.mapToValue(&imap);
+                keys.append(* ((ConnectionId *) imap.getKey()));
+                locks.append(sizeof(LockData), lD);
+                ++nlocks;
+            }
+        }
+
+        if (nlocks < 1)
+            return false;
+
+        if (isFile)
+        {
+            localMB.append(true);
+            localMB.append(file.str());
+        }
+        else
+        {
+            localMB.append(false);
+            localMB.append(xpath);
+        }
+
+        unsigned lockCount=0;
+        size32_t lockCountPos = localMB.length();
+        localMB.append(lockCount);
+
+        unsigned l = 0;
+        unsigned msNow = msTick();
+        loop
+        {
+            LockData lD;
+            memcpy(&lD, ((const byte *)locks.toByteArray())+l*sizeof(LockData), sizeof(LockData));
+
+            StringBuffer sessEpStr;
+            unsigned lockedFor = msNow-lD.timeLockObtained;
+            const char* endpoint = querySessionManager().getClientProcessEndpoint(lD.sessId, sessEpStr).str();
+            if (filters.checkEndpoint(endpoint) && filters.checkDuration(lockedFor))
+            {
+                localMB.append(endpoint);
+                localMB.append(lD.sessId);
+                localMB.append(keys.item(l));
+                localMB.append(lD.mode);
+                localMB.append(lockedFor);
+                lockCount++;
+            }
+            ++l;
+            if (l>=nlocks)
+                break;
+        }
+        if (lockCount < 1)
+            return false;
+
+        localMB.writeDirect(lockCountPos,sizeof(lockCount),&lockCount);
+        mb.append(localMB);
```

Comment:
bit of a waste/inefficient, building up localMB, only to append it into output buffer.
if lockCount == 0, you could reset the mb buffer to retract the xpath you add upfront.